src/uct/ib/mlx5/dv/ib_mlx5dv_md.c

```diff
@@ -2621,85 +2623,87 @@ static ucs_status_t uct_ib_mlx5_devx_xgvmi_umem_mr(uct_ib_mlx5_md_t *md,
-            return status;
-        }
-    } else {
-        status = uct_ib_mlx5_devx_umr_mkey_bind(md, mr, &umr);
-        if (status != UCS_OK) {
-            return status;
-        }
-
-        exported_lkey = umr->umr_mkey->lkey;
+    memh->cross_mr      = cross_mr;
+    memh->exported_lkey = exported_lkey;
+    return UCS_OK;
+}
 
-        /* Return if mkey is already marked as XGVMI */
-        if (uct_ib_mlx5_devx_umr_mkey_is_xgvmi(exported_lkey)) {
-            goto out;
-        }
+static ucs_status_t uct_ib_mlx5_devx_xgvmi_umr_mr(uct_ib_mlx5_md_t *md,
+                                                  uct_ib_mlx5_devx_mem_t *memh)
+{
+    uct_ib_mlx5_devx_umr_mkey_t *umr_mkey =
+        uct_ib_mlx5_devx_umr_mkey_bind(md, &memh->mrs[UCT_IB_MR_DEFAULT]);
+    if (NULL == umr_mkey) {
+        return UCS_ERR_IO_ERROR;
     }
 
-    status = uct_ib_mlx5_devx_allow_xgvmi_access(md, exported_lkey, 1);
-    if (status == UCS_OK) {
-        if (NULL != umr) {
-            /* Mark mkey as XGVMI allowed */
-            exported_lkey = uct_ib_mlx5_devx_umr_mkey_set_xgvmi(umr);
-        }
+    memh->exported_umr_mkey = umr_mkey;
+    memh->exported_lkey     = umr_mkey->mkey->lkey;
+    return UCS_OK;
+}
 
-        goto out;
+UCS_PROFILE_FUNC_ALWAYS(ucs_status_t, uct_ib_mlx5_devx_reg_exported_key,
+                        (md, memh), uct_ib_mlx5_md_t *md,
+                        uct_ib_mlx5_devx_mem_t *memh)
+{
+    size_t length = memh->mrs[UCT_IB_MR_DEFAULT].super.ib->length;
+    ucs_status_t status;
+
+    if (memh->dm != NULL) {
+        return UCS_ERR_UNSUPPORTED;
     }
 
-    ucs_debug("%s: indirect cross gvmi not supported, fallback to DEVX UMEM",
-              uct_ib_device_name(&md->super.dev));
+    ucs_assertv((memh->exported_umr_mkey == NULL) && (memh->cross_mr == NULL),
+                "memh=%p exported_umr_mkey=%p cross_mr=%p exported_lkey=0x%x",
+                memh, memh->exported_umr_mkey, memh->cross_mr,
+                memh->exported_lkey);
 
-    if (cross_mr != NULL) {
-        mlx5dv_devx_obj_destroy(cross_mr);
-    }
+    if (md->flags & UCT_IB_MLX5_MD_FLAG_INDIRECT_XGVMI) {
+        /* UMR bind impl (IBV_WR_BIND_MW) attaches a single KLM segment, so:
+        * - IBV_WR_BIND_MW supports the maximum region length of 2GB
+        * - IBV_WR_BIND_MW does not support multi-segment (multi-threaded) MRs
+        * For these use cases we fallback to KSM */
+        if ((length > UCT_IB_MD_MAX_MR_SIZE) ||
+            (memh->super.flags & UCT_IB_MEM_MULTITHREADED)) {
+            status = uct_ib_mlx5_devx_xgvmi_ksm_mr(md, memh);
+        } else {
+            status = uct_ib_mlx5_devx_xgvmi_umr_mr(md, memh);
+        }
 
-    if (umr != NULL) {
-        uct_ib_mlx5_devx_umr_mkey_destroy(md, umr);
-    }
+        /* If KSM or UMR implementation fail to enable XGVMI, this capability
+         * flag is removed by impl, and then we fallback to UMEM impl */
+        if (md->flags & UCT_IB_MLX5_MD_FLAG_INDIRECT_XGVMI) {
```

Comment:
it seems a bit unexpected check
maybe we could check (status != UCS_ERR_UNSUPPORTED) ?

Reply:
Yes, I have already elaborated this idea.
In short, it's a bit more risky but doable, we just need to be careful, because UCS_ERR_UNSUPPORTED might be returned from other functions (not from "allow xgvmi").
For instance:
```
uct_ib_mlx5_devx_reg_xgvmi_ksm_mr
    uct_ib_mlx5_devx_reg_ksm_data
        uct_ib_mlx5_devx_reg_ksm_data_contig
            uct_ib_mlx5_devx_reg_ksm_data_addr
                uct_ib_mlx5_devx_reg_ksm: line 97: if (mr == NULL) return UCS_ERR_UNSUPPORTED
```
This we should not treat as XGVMI failure, at least it is not treated like that with current impl.
So we can implement it like that:
```
uct_ib_mlx5_devx_reg_xgvmi_ksm_mr()
{
    status = uct_ib_mlx5_devx_reg_ksm_data(); // This may result in UCS_ERR_UNSUPPORTED
    if (status != UCS_OK) {
        return UCS_ERR_IO_ERROR;   << return generic error, lose precise error
    }

    status = uct_ib_mlx5_devx_allow_xgvmi_access(md, exported_lkey, 1);
    if (status != UCS_OK) {
        return UCS_ERR_UNSUPPORTED;
    }
}
```
And then we need to change the `umr_bind` function to return ucs_status_t, and make sure we return UCS_ERR_UNSUPPORTED only in case of xgvmi failure.
It's a bit risky, because some future modification can break this logic. So I'm not sure this is a superior alternative..

Reply:
ok